crates/sparrow-materialize/src/materialize/materialization_control.rs

```diff
@@ -0,0 +1,219 @@
+                    Err(e) => {
+                        self.status =
+                            MaterializationProgress::Failed(ProgressInformation::default(), e);
+                    }
+                }
+            }
+            MaterializationProgress::Running(progress) => {
+                match handle
+                    .await
+                    .into_report()
+                    .change_context(Error::Internal)?
+                    .change_context(Error::Materialization)
+                {
+                    Ok(_) => {
+                        self.status = MaterializationProgress::Stopped(progress.clone());
+                    }
+                    Err(e) => {
+                        self.status = MaterializationProgress::Failed(progress.clone(), e);
+                    }
+                }
+            }
+            MaterializationProgress::Stopped(_) => {
+                tracing::info!("cannot stop; materialization already stopped")
+            }
+            MaterializationProgress::Failed(_, _) => {
+                tracing::info!("cannot stop; materialization already failed")
+            }
+        };
+
+        Ok(())
+    }
+
+    /// Gets the status of the materialization.
+    ///
+    /// Note that progress is reported periodically, and may not contain the exact
+    /// up-to-date numbers.
+    pub fn get_status(&self) -> MaterializationStatus {
+        let progress = self.progress_receiver.borrow();
+        match &*progress {
+            MaterializationProgress::Uninitialized => MaterializationStatus {
+                id: self.id.clone(),
+                progress: ProgressInformation::default(),
+                state: State::Uninitialized,
+                error: None,
+            },
+            MaterializationProgress::Running(progress) => MaterializationStatus {
+                id: self.id.clone(),
+                progress: progress.clone(),
+                state: State::Running,
+                error: None,
+            },
+            MaterializationProgress::Stopped(progress) => MaterializationStatus {
+                id: self.id.clone(),
+                progress: progress.clone(),
+                state: State::Stopped,
+                error: None,
+            },
+            MaterializationProgress::Failed(progress, e) => MaterializationStatus {
+                id: self.id.clone(),
+                progress: progress.clone(),
+                state: State::Failed,
+                error: Some(e.to_string()),
+            },
+        }
+    }
+}
+
+/// Publish status of a materialization.
+#[derive(Debug, Default)]
+pub struct MaterializationStatus {
+    /// Unique identifier of the materialization
+    pub id: String,
+    /// Progress information
+    pub progress: ProgressInformation,
+    /// Current state of the materialization
+    pub state: State,
+    /// Failure reason
+    ///
+    /// Only populated if the materialization has failed
+    pub error: Option<String>,
```

Comment:
Why `String` -- it seems like it would be better for the consumer of this to receive the actual error, and then it can choose to convert to a string if desired. For instance, this discards any information about error codes that may be attached to the error, etc.